hikari/impl/event_manager_base.py

```diff
@@ -50,46 +56,130 @@
+
+    else:
+        event_types = event_types.dispatches()
+
+    def decorator(method: UnboundMethodT[EventManagerBaseT], /) -> UnboundMethodT[EventManagerBaseT]:
+        setattr(method, _CACHE_RESOURCE_ATTRIBUTE, cache_resource)
+        setattr(method, _EVENT_TYPES_ATTRIBUTE, event_types)
+        return method
+
+    return decorator
+
+
+@attr.frozen()
+class _Consumer:
+    callback: ConsumerT
+    cache: undefined.UndefinedOr[config.CacheComponents]
+    event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]
+
+
 class EventManagerBase(event_manager.EventManager):
     """Provides functionality to consume and dispatch events.
 
     Specific event handlers should be in functions named `on_xxx` where `xxx`
     is the raw event name being dispatched in lower-case.
     """
 
-    __slots__: typing.Sequence[str] = ("_app", "_listeners", "_consumers", "_waiters")
+    __slots__: typing.Sequence[str] = ("_app", "_dispatches_for_cache", "_listeners", "_consumers", "_waiters")
 
     def __init__(self, app: traits.BotAware) -> None:
         self._app = app
-        self._consumers: typing.Dict[str, ConsumerT] = {}
+        self._dispatches_for_cache: typing.Dict[_Consumer, bool] = {}
+        self._consumers: typing.Dict[str, _Consumer] = {}
         self._listeners: ListenerMapT[base_events.Event] = {}
         self._waiters: WaiterMapT[base_events.Event] = {}
 
         for name, member in inspect.getmembers(self):
             if name.startswith("on_"):
-                self._consumers[name[3:]] = member
+                member = typing.cast("MethodT", member)
+                cache_resource = getattr(member, _CACHE_RESOURCE_ATTRIBUTE, undefined.UNDEFINED)
+                event_types = getattr(member, _EVENT_TYPES_ATTRIBUTE, undefined.UNDEFINED)
+                cache_resource = typing.cast("undefined.UndefinedOr[config.CacheComponents]", cache_resource)
+                event_types = typing.cast(
+                    "undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]", event_types
+                )
+                self._consumers[name[3:]] = _Consumer(member, cache_resource, event_types)
+
+    def _enabled_for_event(self, event_type: typing.Type[base_events.Event], /) -> bool:
+        for cls in event_type.dispatches():
+            if cls in self._listeners or cls in self._waiters:
+                return True
+
+        return False
+
+    # This returns int rather than bool to avoid unnecessary bool casts
+    def _enabled_for_consumer(self, consumer: _Consumer) -> int:
+        # If undefined then we can only safely assume that this does link to registered listeners.
+        if consumer.event_types is undefined.UNDEFINED:
+            return True
+
+        if (cached_value := self._dispatches_for_cache.get(consumer, ...)) is True:
+            return True
+
+        if cached_value is ...:
```

Comment:
if it's not cached (`...`) then we want to calculate the actual value, otherwise the cached value is `False` and we want to skip straight ahead to the cache components check

Reply:
Cached value can only ever be a bool tho :thinking:

Reply:
you're missing the default in `self._dispatches_for_cache.get(consumer, ...)`

Reply:
Ik, which is why I'm saying there is no need for it. It could just be `self._dispatches_for_cache.get(consumer)` and you use the default as `None`, no?

Reply:
Either works, they're both standard singletons

Reply:
I just mean because it will make it easier to read